### PR TITLE
Implementations

### PR DESCRIPTION
**Pull request by Grayahm**
Implemented:

- Snowpoint Temple
- Azelf Lv.X
- Mesprit Lv.X
- Regigigas Lv.X
- Dialga G
- Dialga G Lv.X

### CODE REVIEW

File: src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy [hunk 1]
```diff
@@ -3112,15 +3125,18 @@ public enum LegendsAwakened implements LogicCardInfo {
             energyCost ()
             attackRequirement {}
```
Comment (flagrama) [inline src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy:3126]: I'd probably put a `assert my.bench` here as without a benched Pokémon this attack would have no effect.

File: src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy [hunk 2]
```diff
@@ -3112,15 +3125,18 @@ public enum LegendsAwakened implements LogicCardInfo {
...
           move "Supreme Blast", {
-            text "200 damage. in play, this attack does nothing. Discard all Energy attached to Mesprit."
+            text "200 damage. If you don't have Uxie LV.X and Azelf LV.X in play, this attack does nothing. Discard all Energy attached to Mesprit."
             energyCost P, P
-            attackRequirement {}
+            attackRequirement {my.all.findAll{it.name == ("Uxie Lv.X")} && my.all.findAll{it.name == ("Uxie Lv.X")}}
             onAttack {
-              damage 0
+              damage 200
+              discardAllEnergySelf(null)
```
Comment (flagrama) [inline src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy:3139]: The method is `discardAllSelfEnergy()` and you don't need to include the `null` as it has null as the default value.

File: src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy [hunk 3]
```diff
@@ -3195,14 +3211,22 @@ public enum LegendsAwakened implements LogicCardInfo {
           pokePower "Trade Off", {
             text "Once during your turn , you may look at the top 2 cards of your deck, choose 1 of them, and put it into your hand. Put the other card on the bottom of your deck. This power can’t be used if Uxie is affected by a Special Condition. You can’t use more than 1 Trade Off Poké-Power each turn."
             actionA {
+              checkNoSPC()
+              checkLastTurn()
+              assert my.deck : "Your deck is empty"
+              powerUsed()
+              def sel = my.deck.subList(0,2).select("Choose 1 card to put into your hand")
+              my.deck.subList(0,2).getExcludedList(sel).moveTo(suppressLog: true, my.deck)
```
Comment (flagrama) [inline src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy:3219]: I don't see `suppressLog` anywhere else in the codebase. Are you sure `hidden` won't do what you want here like it does when moving to your hand?

File: src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy [hunk 4]
```diff
@@ -3195,14 +3211,22 @@ public enum LegendsAwakened implements LogicCardInfo {
           pokePower "Trade Off", {
```
Comment (flagrama) [inline src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy:3211]: This power isn't fully implemented. You can't use more than 1 Trade Off per turn, so you have to make sure other Pokémon with the same power can't use it. Check the Aegislash in RebelClash for an example on bwAbility.

File: src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy [hunk 5]
```diff
@@ -3195,14 +3211,22 @@ public enum LegendsAwakened implements LogicCardInfo {
...
             }
           }
           move "Zen Blade", {
             text "60 damage. Uxie can’t use Zen Blade during your next turn."
             energyCost C, C
             attackRequirement {}
             onAttack {
-              damage 0
+              damage 60
+              cantAttackNextTurn self
```
Comment (flagrama) [inline src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy:3229]: According to the text, Uxie can still attack, it just can't use this same attack.
```suggestion
              cantUseAttack thisMove, self
```

File: src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy (hunk 3 above)
Comment (Grayahm) [reply src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy:3219]: As in `.moveTo(hidden, my.deck)` ?

File: src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy (hunk 4 above)
Comment (Grayahm) [reply src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy:3212]: Pretty sure I fixed this correctly now

Comment (flagrama) [reply src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy:3212]: Yeah, I think it looks good now.

File: src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy [hunk 6]
```diff
@@ -3112,15 +3125,19 @@ public enum LegendsAwakened implements LogicCardInfo {
...
           move "Supreme Blast", {
-            text "200 damage. in play, this attack does nothing. Discard all Energy attached to Mesprit."
+            text "200 damage. If you don't have Uxie LV.X and Azelf LV.X in play, this attack does nothing. Discard all Energy attached to Mesprit."
             energyCost P, P
-            attackRequirement {}
+            attackRequirement {my.all.findAll{it.name == ("Uxie Lv.X")} && my.all.findAll{it.name == ("Uxie Lv.X")}}
             onAttack {
-              damage 0
+              damage 200
+              discardAllEnergySelf()
```
Comment (flagrama) [inline src/tcgwars/logic/impl/gen4/LegendsAwakened.groovy:3140]: The order of Energy and Self is still incorrect. `discardAllEnergySelf` may make more sense using english rules, but `discardAllSelfEnergy` is actually in the codebase.

**Comment by Grayahm** [comment]
Think everything is fixed

**Pull request merged**